src/directory/footer.rs

```diff
@@ -86,16 +89,41 @@ impl Footer {
 
     /// Confirms that the index will be read correctly by this version of tantivy
     /// Has to be called after `extract_footer` to make sure it's not accessing uninitialised memory
-    pub fn is_compatible(&self) -> bool {
-        let version = &*crate::VERSION;
+    pub fn is_compatible(&self) -> Result<bool, TantivyError> {
```

Comment:
I don't understand the dichotomy between incompatibility with false and incompatibiliy expressed by an error.
What is the point?

Reply:
I thought adding an error type makes it easier to surface it the error type to the top, where bool will fail without any information on how to fix it.

Reply:
This is way too complicated for no reason at all.

Let's have it return `Result<(), IncompatibleIndex>` or `bool`.

Reply:
Returning an Ok(Incompatibility::CompressionMismatch) feels semantically weird to me. 

Index incompatiblity is an error after which we cannot do anything else with the index. 

> Also what is the point of the intermediary `IncompatibleIndex` ?

I don't understand your usage of the word intermediary - can you please explain

Reply:
| Returning an Ok(Incompatibility::CompressionMismatch) feels semantically weird to me.

I think it would be clearer.

My bad I think I misunderstood the nature of `IncompatibleIndex`. I think you can use `Incompatibility` as an error directly though.

Reply:
If I understand your suggestion, the caller of is_compatible will unwrap the always Ok(incompatibility) and pattern match against different variants.  

Surely, we then lose the user friendly errors defined in TantivyError:: incompatible?

Reply:
Err(Incompatibility) you mean? Why do we lose the user friendly errors?